src/resolve-info.ts

```diff
@@ -0,0 +1,247 @@
+type GraphQLNamedOutputType = GraphQLNamedType & GraphQLOutputType;
+type GraphQLObjectLike = GraphQLInterfaceType | GraphQLObjectType;
+
+const memoizedGetReturnType = memoize2(getReturnType);
+const memoizedHasField = memoize2(hasField);
+const memoizedResolveEndType = memoize(resolveEndType);
+const memoizedGetPossibleTypes = memoize2(getPossibleTypes);
+const memoizedExpandFieldNodeType = memoize4(expandFieldNodeType);
+const memoizedExpandFieldNode = memoize4(expandFieldNode);
+
+function expandFieldNode(
+  schema: GraphQLSchema,
+  fragments: FragmentsType,
+  node: FieldNode,
+  fieldType: GraphQLOutputType
+): FieldExpansion | true {
+  if (node.selectionSet == null) {
+    return true;
+  }
+
+  // there is a selectionSet which makes the fieldType a CompositeType
+  const typ = memoizedResolveEndType(fieldType) as GraphQLCompositeType;
+  const possibleTypes = memoizedGetPossibleTypes(schema)(typ);
+
+  const fieldExpansion: FieldExpansion = {};
+  for (const possibleType of possibleTypes) {
+    if (!isUnionType(possibleType)) {
+      fieldExpansion[possibleType.name] = memoizedExpandFieldNodeType(schema)(
+        fragments
+      )(possibleType)(node.selectionSet);
+    }
+  }
+
+  return fieldExpansion;
+}
+
+function expandFieldNodeType(
+  schema: GraphQLSchema,
+  fragments: FragmentsType,
+  parentType: GraphQLCompositeType,
+  selectionSet: SelectionSetNode
+): TypeExpansion {
+  const typeExpansion: TypeExpansion = {};
+
+  for (const selection of selectionSet.selections) {
+    if (selection.kind === "Field") {
+      if (
+        !isUnionType(parentType) &&
+        memoizedHasField(parentType)(selection.name.value)
+      ) {
+        typeExpansion[selection.name.value] = memoizedExpandFieldNode(schema)(
+          fragments
+        )(selection)(memoizedGetReturnType(parentType)(selection.name.value));
+      }
+    } else {
+      const selectionSet =
+        selection.kind === "InlineFragment"
+          ? selection.selectionSet
+          : fragments[selection.name.value].selectionSet;
+      deepMerge(
+        typeExpansion,
+        memoizedExpandFieldNodeType(schema)(fragments)(parentType)(selectionSet)
+      );
+    }
+  }
+
+  return typeExpansion;
+}
+
+/**
+ * Returns a list of Possible types that one can get to from the
+ * resolvedType. As an analogy, these are the same types that one
+ * can use in a fragment's typeCondition.
+ *
+ * Note: This is different from schema.getPossibleTypes() that this
+ * returns all possible types and not just the ones from the type definition.
+ *
+ * Example:
+ * interface Node {
+ *   id: ID!
+ * }
+ * type User implements Node {
+ *   id: ID!
+ *   name: String
+ * }
+ * type Article implements Node {
+ *   id: ID!
+ *   title: String
+ * }
+ * union Card = User | Article
+ *
+ * - schema.getPossibleTypes(Card) would give [User, Article]
+ * - This function getPossibleTypes(schema, Card) would give [User, Article, Node]
```

Comment:
at the resolution level, we do not know in advance the resolved type already. But we might know about one of the overlapping interfaces and know at the time of resolution that it will be of that type. For example, a union of types that share a common interface.

Reply:
makes sense.